Add support for test/retest command in pull request comment

glefloch opened the pull request:
This branch aims to add commands handling on pull request command. 

On a pull request comment, it :
* check this is a comment on a pull request
* the user that commented this is not `@quarkus-bot`
* validate the user is part the `quarkus-push` teams

Once it is ok, it looks for a `Command` bean that match the comment body. 

The `RerunWorkflowCommand` accepts `test` and `retest` allowing: 
* `@quarkus-bot test` command 
* `@quarkus-bot retest` command

Regarding the workflow, it looks for **Quarkus CI** `WorkflowRun`, take the last one, and `rerun` it. 

I created the PR in draft as I'm not able to get a working `rerun`, no error, is thrown by the API. When I log the url of the workflow that should be rerun, it is the correct one. This may come from my repo not providing enough permissions but I'm not able to find what's missing, maybe @gsmet you will have some hints?

glefloch committed:
Add support for test/restest command in pull request comment

glefloch commented on src/main/java/io/quarkus/bot/PullRequestCommandHandler.java:
```diff
@@ -0,0 +1,70 @@
+    private static final Logger LOG = Logger.getLogger(PullRequestCommandHandler.class);
+
+    private static final String QUARKUS_BOT_NAME = "quarkus-bot";
+    private static final String QUARKUS_TEAM_MEMBERS = "quarkus-push";
```
This could be externalized to ease development on test repo

gsmet commented on src/main/java/io/quarkus/bot/PullRequestCommandHandler.java:
```diff
@@ -0,0 +1,69 @@
+
+    private static final Logger LOG = Logger.getLogger(PullRequestCommandHandler.class);
+
+    private static final String QUARKUS_BOT_NAME = "quarkus-bot";
```
Are you sure this will be the name of the bot? IIRC, GH adds something like `[bot]` after it.

glefloch replied:
yes, you are right, I updated it.

gsmet commented on src/main/java/io/quarkus/bot/PullRequestCommandHandler.java:
```diff
@@ -0,0 +1,69 @@
+        GHIssue issue = commentPayload.getIssue();
+
+        if (issue.isPullRequest()
+                && !QUARKUS_BOT_NAME.equals(user.getName())) {
```
I usually recommend to do the opposite: test the negative and evacuate this case with a return;. That way you keep your logic very simple with no nested ifs.

gsmet replied:
I think I would use `getUserName()` instead even if deprecated as it should avoid one API call. Keep in mind that this will be called a lot so better avoid doing one API call per added comment.

gsmet commented on src/main/java/io/quarkus/bot/PullRequestCommandHandler.java:
```diff
@@ -0,0 +1,69 @@
+                if (!quarkusBotConfig.isDryRun()) {
+                    ghIssueConsumer.get().run(issue);
+                } else {
+                    LOG.info("Pull request #" + issue.getNumber() + " - Restart workflow.");
```
I think you shouldn't do this here but it's the action that should be executed and the log should be added there.

gsmet commented on src/main/java/io/quarkus/bot/PullRequestCommandHandler.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    private boolean isQuarkusTeamMember(GHUser user, GHOrganization organization) throws IOException {
+        GHTeam quarkusPushTeam = organization.getTeams().get(QUARKUS_TEAM_MEMBERS);
```
I think it would be better to check the `write` permission with `GHRepository.getPermission(GHUser)`.

gsmet commented on src/main/java/io/quarkus/bot/command/RerunWorkflowCommand.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.Optional;
+
+@ApplicationScoped
+public class RerunWorkflowCommand implements Command {
```
Why don't you make `Command` parameterized?

gsmet commented on src/main/java/io/quarkus/bot/command/RerunWorkflowCommand.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @Override
+    public void run(GHIssue issue) {
```
And here I would pass the pull request rather than the issue.

gsmet commented on src/main/java/io/quarkus/bot/command/RerunWorkflowCommand.java:
```diff
@@ -0,0 +1,56 @@
+                    .list().toList();
+
+            Optional<GHWorkflowRun> ghWorkflowRun = ghWorkflowRuns.stream()
+                    .filter(workflowRun -> WorkflowConstants.QUARKUS_CI_WORKFLOW_NAME.equals(workflowRun.getName()))
```
I think you also need to rerun the documentation workflow if there.

gsmet commented on src/main/java/io/quarkus/bot/command/RerunWorkflowCommand.java:
```diff
@@ -0,0 +1,56 @@
+                LOG.warn("Unable to find a workflow run to rerun");
+            }
+        } catch (IOException e) {
+            LOG.error("Unable to run workflow", e);
```
Just throw the exception, the framework will take care of it.

gsmet commented on src/main/java/io/quarkus/bot/command/RerunWorkflowCommand.java:
```diff
@@ -0,0 +1,56 @@
+                LOG.info("Rerunning workflow: " + ghWorkflowRun.get().getHtmlUrl());
+                ghWorkflowRun.get().rerun();
+            } else {
+                LOG.warn("Unable to find a workflow run to rerun");
```
I think you should use the reaction feature here, similarly to what Dependabot does with the commands (i.e. add a reaction to the comment). See `GHIssueComment#createReaction()`.

You would either have to push the issue comment here or you could return a boolean (or a `ReactionContent`) from the run method and centralize it in the other class. I think the latter would probably be better.

gsmet commented on src/main/java/io/quarkus/bot/PullRequestCommandHandler.java:
```diff
@@ -0,0 +1,69 @@
+        if (issue.isPullRequest()
+                && !QUARKUS_BOT_NAME.equals(user.getName())) {
+            Optional<Command> ghIssueConsumer = extractCommand(commentPayload.getComment().getBody());
+            if (ghIssueConsumer.isPresent() && isQuarkusTeamMember(user, commentPayload.getOrganization())) {
```
Yeah good thing to check the permission second to limit the number of API calls.

gsmet reviewed (changes requested):
Cool! I added a bunch of comments, mostly about details.

As for the rerunning the workflow, AFAIK the API silently ignore these calls if you don't have the permissions. I got hit by that too.

You need to go to this page (well the one of your instance): https://github.com/organizations/quarkusio/settings/apps/quarkus-bot/permissions and make sure you have `Read & Write` for `Actions`. Also, you will receive an email asking you to confirm the new privileges. If you don't do it, they won't be taken into account.

gsmet committed:
Simplify logic of PullRequestCommandHandler

In passing, only post a reaction if a command has been detected.

gsmet committed:
Only return rocket reaction if a workflow run is restarted

gsmet committed:
Normalize the executed command

gsmet committed:
Only call getPermission() once to limit the number of API calls

gsmet committed:
Use a personal access token to trigger the workflow rerun

There is a bug in the GitHub API and for now it's not possible to
trigger a rerun with an installation token. We have no ETA for the fix
so using a personal access token is the only workaround for now.
